common/crypto/curl-p/ptrit.c

```diff
@@ -1,75 +1,323 @@
+  }
+
+  for (i = 0; i < CURL_STATE_SIZE / 4; ++i) {
+    pcurl_s2(b++, c, bb++);
+    pcurl_s2(c++, b, bb++);
+  }
+#elif defined(PCURL_SBOX_UNWIND_4)
+  // 728 = 8*91
+  for (i = 0; i < CURL_STATE_SIZE / 8; ++i) {
+    pcurl_s2(b++, c, a++);
+    pcurl_s2(c++, b, a++);
+    pcurl_s2(b++, c, a++);
+    pcurl_s2(c++, b, a++);
   }
+
+  for (i = 0; i < CURL_STATE_SIZE / 8; ++i) {
+    pcurl_s2(b++, c, bb++);
+    pcurl_s2(c++, b, bb++);
+    pcurl_s2(b++, c, bb++);
+    pcurl_s2(c++, b, bb++);
+  }
+#else
+#error Invalid PCURL_SBOX_UNWIND.
+#endif
+
+  *bb = *a0;
 }
+#else
+#error Invalid PCURL_STATE.
+#endif  // PCURL_STATE
 
-void ptrit_curl_squeeze(PCurl *const ctx, ptrit_t *const trits, size_t length) {
-  size_t num_chunks = length / HASH_LENGTH_TRIT + ((length % HASH_LENGTH_TRIT) ? 1 : 0);
-  size_t i = 0;
+void pcurl_init(pcurl_t *ctx, size_t round_count) {
+  pcurl_reset(ctx);
+  ctx->round_count = round_count;
+}
+void pcurl_absorb(pcurl_t *ctx, ptrit_t const *ptrits, size_t length) {
+  size_t n = length / CURL_RATE;
+  for (; n--;) {
+    memcpy(ctx->state, ptrits, CURL_RATE * sizeof(ptrit_t));
+    ptrits += CURL_RATE;
+    // no padding!
+    pcurl_transform(ctx);
+  }
+  length %= CURL_RATE;
+
+  if (0 < length) {
+    memcpy(ctx->state, ptrits, length * sizeof(ptrit_t));
+    ptrits += length;
+    // no padding!
+    pcurl_transform(ctx);
+    length = 0;
+  }
+}
+void pcurl_squeeze(pcurl_t *ctx, ptrit_t *ptrits, size_t length) {
+  size_t n = length / CURL_RATE;
+  for (; n--;) {
+    memcpy(ptrits, ctx->state, CURL_RATE * sizeof(ptrit_t));
+    ptrits += CURL_RATE;
+    // no padding!
+    pcurl_transform(ctx);
+  }
+  length %= CURL_RATE;
 
-  for (; i < num_chunks; ++i) {
-    memcpy(trits + i * HASH_LENGTH_TRIT, ctx->state,
-           (length < HASH_LENGTH_TRIT ? length : HASH_LENGTH_TRIT) * sizeof(ptrit_t));
-    ptrit_transform(ctx);
-    length = length < HASH_LENGTH_TRIT ? 0 : length - HASH_LENGTH_TRIT;
+  if (0 < length) {
+    memcpy(ptrits, ctx->state, length * sizeof(ptrit_t));
+    ptrits += length;
+    // no padding!
+    pcurl_transform(ctx);
+    length = 0;
   }
 }
+void pcurl_get_hash(pcurl_t *ctx, ptrit_t *hash) {
+  memcpy(hash, ctx->state, CURL_RATE * sizeof(ptrit_t));
+  pcurl_reset(ctx);
+}
+void pcurl_hash_data(pcurl_t *ctx, ptrit_t const *data, size_t size, ptrit_t *hash) {
+  pcurl_reset(ctx);
+  pcurl_absorb(ctx, data, size);
+  pcurl_get_hash(ctx, hash);
+}
 
-void ptrit_transform(PCurl *const ctx) {
-  PCurl s;
-  size_t round = 0;
-  ptrit_t *lhs, *rhs;
+#if defined(PCURL_STATE_SHORT)
+void pcurl_transform(pcurl_t *ctx) {
+  size_t round;
 
-  for (; round < ctx->type; ++round) {
-    if (round & 1) {
-      lhs = ctx->state;
-      rhs = s.state;
-    } else {
-      lhs = s.state;
-      rhs = ctx->state;
-    }
-    ptrit_sbox(lhs, rhs);
+  ptrit_t *a = ctx->state;
+  ptrit_t *b = a + (CURL_STATE_SIZE + 1) / 2;
+  ptrit_t *c = b + (CURL_STATE_SIZE + 1) / 2;
+  b[364] = a[0];
```

Comment:
Should we use a constant here and everywhere we use this ?

Reply:
Could you please clarify? Not sure what you mean.

Reply:
`#define CURL_STATE_HALF_SIZE ((CURL_STATE_SIZE + 1) / 2)`
364 = `CURL_STATE_HALF_SIZE - 1`
Done.